controllers/bookController.go

```diff
@@ -0,0 +1,219 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/avigailbr/book_service/config"
+	"github.com/avigailbr/book_service/connectors"
+	"github.com/avigailbr/book_service/models"
+	"gopkg.in/olivere/elastic.v5"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+var (
+	BookController IBookController
+)
+
+type IBookController interface {
+	GetBook(id string) (*models.Book, error)
+	InsertBook(book *models.Book) (string, error)
+	UpdateBook(id string, fields map[string]interface{}) error
+	DeleteBook(id string) error
+	Search(fields map[string]string) ([]models.Book, error)
+	Info() (map[string]interface{}, error)
+}
+
+type ElasticBookController struct {
+}
+
+func NewElasticBookController() IBookController {
+	return &ElasticBookController{}
+}
+
+func (e ElasticBookController) GetBook(id string) (*models.Book, error) {
+	es := connectors.BookConnectors.ElasticClient
+	ctx := context.Background()
+	get, err := es.Client.Get().
+		Index(es.Index).
+		Type(es.DocType).
+		Id(id).
+		Do(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var book models.Book
+	if data, err := json.Marshal(get.Source); err != nil {
+		return nil, err
+	} else {
+		json.Unmarshal(data, &book)
+		return &book, nil
+	}
+
+}
+
+func (e ElasticBookController) InsertBook(book *models.Book) (string, error) {
+	es := connectors.BookConnectors.ElasticClient
+	ctx := context.Background()
+
+	data, err := json.Marshal(book)
+	if err != nil {
+		return "", err
+	}
+
+	put, err := es.Client.Index().
+		Index(es.Index).
+		Type(es.DocType).
+		BodyJson(string(data)).
+		Do(ctx)
+
+	if err != nil {
+		return "", err
+	}
+
+	return put.Id, nil
+
+}
+
+func (e ElasticBookController) UpdateBook(id string, fields map[string]interface{}) error {
+	es := connectors.BookConnectors.ElasticClient
+	ctx := context.Background()
+
+	_, err := es.Client.Update().
+		Index(es.Index).
+		Type(es.DocType).
+		Id(id).
+		Doc(fields).
+		Do(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e ElasticBookController) DeleteBook(id string) error {
+	es := connectors.BookConnectors.ElasticClient
+	// must pass a context to execute each service
+	ctx := context.Background()
+	_, err := es.Client.Delete().
+		Index(es.Index).
+		Type(es.DocType).
+		Id(id).
+		Do(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e ElasticBookController) Search(fields map[string]string) ([]models.Book, error) {
+
+	query := elastic.NewBoolQuery()
+
+	if title, ok := fields["title"]; ok && title != "" {
+		titleMatchQuery := elastic.NewMatchQuery("title", title)
+
+		if author, ok := fields["author_name"]; ok && author != "" {
+			authorMatchQuery := elastic.NewMatchQuery("author_name", author)
+			query.Must(titleMatchQuery, authorMatchQuery)
+
+		} else {
+			query.Must(titleMatchQuery)
+		}
+
+	} else if author, ok := fields["author_name"]; ok && author != "" {
+		authorMatchQuery := elastic.NewMatchQuery("author_name", author)
+		query.Must(authorMatchQuery)
+	}
```

Comment:
you can just use an if for each param, instead of the nesting..
```suggestion
	if title, ok := fields["title"]; ok && title != "" {
		titleMatchQuery := elastic.NewMatchQuery("title", title)
		query.Must(titleMatchQuery)
	}
	if author, ok := fields["author_name"]; ok && author != "" {
		authorMatchQuery := elastic.NewMatchQuery("author_name", author)
		query.Must(authorMatchQuery)
	}
```